MatrixRTC: New membership manager

toger5 opened the pull request:
Fixes: https://github.com/element-hq/voip-internal/issues/309
Part of https://github.com/element-hq/element-call/issues/2972.

This does not yet use the new manager by default. It only runs the test suite with it.

Notes on error cases encountered when making HTTP requests to the Homeserver:

Type | `MatrixClient` stack | `RoomWidgetClient` stack
-|-|-
"Local" request timeout enforced by the Matrix client | [throws](https://github.com/matrix-org/matrix-js-sdk/blob/dabb8aab22083eb30b928ed53d6493a84acf6593/src/http-api/fetch.ts#L330-L332) `Error` with `name` === `AbortError` | ?
HTTP 5xx response from homeserver | [throws](https://github.com/matrix-org/matrix-js-sdk/blob/dabb8aab22083eb30b928ed53d6493a84acf6593/src/http-api/fetch.ts#L339) [`HttpError`](https://github.com/matrix-org/matrix-js-sdk/blob/71bffb6c1bb89b5d9802667bee2626e29a4f57d5/src/http-api/utils.ts#L103-L106) | throws `MatrixError`
Network/socket error/timeout | [throws](https://github.com/matrix-org/matrix-js-sdk/blob/dabb8aab22083eb30b928ed53d6493a84acf6593/src/http-api/fetch.ts#L333) `ConnectionError` | ? 
Request timeout between widget and host Matrix client | n/a | now mapped to `ConnectionError` as part of this PR

<!-- Thanks for submitting a PR! Please ensure the following requirements are met in order for us to review your PR -->

## Checklist

- [x] Tests written for new code (and old code if feasible).
- [x] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
- [x] Linter and other CI checks pass.
- [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md)).

hughns reviewed (commented):
Some minor tidying things for now.

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,808 @@
+        }
+    }
+
+    public async membershipLoopHandler(state: ActionSchedulerState, type: MembershipActionType): Promise<void> {
```
`public`? surely not?

hughns replied:
Wrap it like this?

```ts
class ActionScheduler {
    ...
    public constructor(
        state: ActionSchedulerState,
        private membershipLoopHandler: (state: ActionSchedulerState, type: MembershipActionType) => Promise<void>,
    ) {
        this.state = state;
    }
	...
    public async startWithActions(initialActions: Action[]): Promise<void> {
		...
                try {
                    await this.membershipLoopHandler(this.state, nextAction.type as MembershipActionType);
                } catch (e) {
                    throw Error("The MemberhsipManager has to shut down because of the end condition: " + e);
                }
		...
    }
	...

}
...
export class MembershipManager implements IMembershipManager {
	...
	// Scheduler:
    private scheduler = new ActionScheduler(
        {
            hasMemberStateEvent: false,
            running: false,
            nextRelativeExpiry: this.membershipEventExpiryTimeout,
            delayId: undefined,
            rateLimitRetries: 0,
            retries: 0,
        },
        (state, type) => this.membershipLoopHandler(state, type),
    );

    // Loop Handler:
    private async membershipLoopHandler(state: ActionSchedulerState, type: MembershipActionType): Promise<void> {
	...
}
```

toger5 replied:
Oh my response got lost here.
It was some talking about how I dont like binding to this and how using a callback constructor param is a bit of a footgun in js so I try to not use the pattern.
```
scheduler = new ActionScheduler(
        {
            hasMemberStateEvent: false,
            running: false,
            nextRelativeExpiry: this.membershipEventExpiryTimeout,
            delayId: undefined,
            rateLimitRetries: 0,
            retries: 0,
        },
       // (state, type) => this.membershipLoopHandler(state, type),
       // this is the footgun its easy to assume this should work but then the wrong this is bound...
        this.membershipLoopHandler,
    );
```

But then I wanted to try it anyways since the public `membershipLoopHandler` is also bad.

I convinced myself that this is fine, since we will not use this Constructor all over the place (e.g. this is the only place i expect it to be used) Hence noone hopefully runs into the js footgun.

Updated it to:
```
    private scheduler = new ActionScheduler(ActionScheduler.defaultState, this.membershipLoopHandler.bind(this));

```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,808 @@
+    }
+
+    public async membershipLoopHandler(state: ActionSchedulerState, type: MembershipActionType): Promise<void> {
+        switch (type) {
```
I would find it slightly easier to read this if it followed the lifecycle ordering: `SendJoinEvent`, `Update`, `SendLeaveEvent`

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,808 @@
+                // This is only a fallback in case we do not have working delayed events support.
+                // first we should try to just send the scheduled leave event
+                try {
+                    this.client.sendStateEvent(
```
Missing `await`?

toger5 replied:
I will look into the future lint pr today!

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,808 @@
+                        this.room.roomId,
+                        EventType.GroupCallMemberPrefix,
+                        {},
+                        this.makeMembershipStateKey(this.userId, this.deviceId),
```
Can this be `this.stateKey` for consistency?

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,808 @@
+                        await this.manager.membershipLoopHandler(this.state, nextAction.type);
+                    }
+                } catch (e) {
+                    throw Error("The MemberhsipManager has to shut down because of the end condition: " + e);
```
A few thoughts;

- I'm surprised we aren't doing any cleanup of the membership state
- If the intention is that this will only ever be reached when an explicit "end condition" throw was made, then should we be checking the type of error and only terminate if it was of that type
- In the current implementations of the loop handlers there aren't general purpose catch blocks, so I think any unanticipated error would hit this catch block

toger5 replied:
There now is `scheduler.resetState()`

hughns replied:
```suggestion
                    throw Error("The MembershipManager has to shut down because of the end condition: " + e);
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,808 @@
+                this.resetWith = undefined;
+            }
+
+            this.actions = this.actions.filter((a) => a !== nextAction);
```
For some reason I had expected that you would I had expected that you would have popped the original action out of the array and then you wouldn't have to filter here.  🤷‍♂️

toger5 replied:
TLDR: I require the following order: get first element, read this element from the array where i cannot be removed in anther codeblock, then it needs to be removed.
pop makes this atomic but it is required to be two steps.

There are of course workarounds like:
```
this.currentItem = this.array.pop()
getArrayWithCurrent() return [...this.array, this.currentItem]
```
And then only use getArrayWithCurrent but I did not prefer any of the alternatives to what it is now.

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,808 @@
+                        // Success we reset retires and set delayId.
+                        state.sendDelayedEventRetries = 0;
+                        state.delayId = response.delay_id;
+                        this.scheduler.addAction({ ts: Date.now(), type: MembershipActionType.SendJoinEvent });
```
I *think* this is the only place where the `SendJoinEvent` action is scheduled. So, what happens if the server doesn't support the delayed events feature? At the moment I don't *think* `SendJoinEvent` gets run.

toger5 replied:
yes I really need to go through all catch blocks and think about the most reasonable "default" failure mode.
Asking the question: "Can we do the same thing in ALL other failure cases and what would that message be?"
I think that will expose a couple of other specific errors we need to handle specifically and what we need to do in the "default" case.

toger5 replied:
I went through all the catch blocks and opted for throwing in a lot of cases.
Only on the actual event sending I opted for a blind "retry" even though we do not understand the error.
But this blind retry is now also rate limited.

toger5 replied:
This got another update in 5a680b0e9fa43024d93fbad9f675d44b6c0585ac where I added tests for those cases

hughns reviewed (changes requested):
I am finding the revised catch block structure much more readable 👍 

A few things:

1. I think chunks like this are misleading as the `throw` is unreachable:

```ts
// Event sending retry (different to rate limit retries)
if (this.retryOnAnyErrorHandler(e, type)) break;

throw Error("Could not send state event because of unrecoverable error: " + e);
```

I think something like this would be clearer:

```ts
// Event sending retry (different to rate limit retries)
// throws once retry count is reached
this.retryOnAnyErrorHandler(e, type);
```

2. I saw you changed the scheduler implementation to use `pop()` and then back again. Did it not work out?

3. For some actions network and server errors (e.g. socket level timeouts, and HTTP level) will be handled by the `retryOnAnyErrorHandler()` but there are still some actions where we don't:
 - `MembershipActionType.SendMainDelayedEvent` - we throw a fatal Error
 - `MembershipActionType.RestartDelayedEvent` - we throw a fatal Error
 - `MembershipActionType.SendFirstDelayedEvent` - we silently fail (I've suggested adding some logging at a minimum)

I would like to see these handled.

4. For general debugging of this, should we be logging the errors at the start of each `catch` block?

hughns commented on spec/unit/matrixrtc/MembershipManager.spec.ts:
```diff
@@ -338,9 +339,11 @@ describe.each([
             const manager = new TestMembershipManager({}, room, client, () => undefined);
-            manager.leave();
+            const errorFn = jest.fn();
+            await manager.leave().catch(errorFn);
```
I think you can do something like this instead:

```
expect(await manager.leave()).not.toThrow()
```

hughns commented on spec/unit/matrixrtc/MembershipManager.spec.ts:
```diff
@@ -648,13 +647,34 @@ describe.each([
+            expect(unrecoverableError).not.toHaveBeenCalledWith();
+            expect(client.sendStateEvent).toHaveBeenCalled();
+        });
+        it("UnsupportedEndpointError does not in an unrecoverable error. We use the manager without delayed events !FailsForLegacy", async () => {
```
```suggestion
        it("falls back to using pure state events when UnsupportedEndpointError encountered for delayed events !FailsForLegacy", async () => {
```

hughns commented on spec/unit/matrixrtc/MembershipManager.spec.ts:
```diff
@@ -648,13 +647,34 @@ describe.each([
             expect(delayEventRestartError).toHaveBeenCalled();
-        }, 5000);
+        });
+        it("Errors while sending delayed events dont result in an unrecoverable error. We use the manager without delayed events !FailsForLegacy", async () => {
```
```suggestion
        it("falls back to using pure state events when some error occurs while sending delayed events !FailsForLegacy", async () => {
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,859 @@
+import { isLivekitFocusActive } from "./LivekitFocus.ts";
+import { type MembershipConfig } from "./MatrixRTCSession.ts";
+
+const logger = rootLogger.getChild("MatrixRTCSessionManager");
```
My bad, I realise that this is part of MatrixRTCSession not MatrixRTCSessionManager:

```suggestion
const logger = rootLogger.getChild("MatrixRTCSession");
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,859 @@
+                            this.stateKey,
+                        )
+                        .then((response) => {
+                            // Success we reset retires and set delayId.
```
```suggestion
                            // Success we reset retries and set delayId.
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,859 @@
+                        });
+                    // On any other error we fall back to not using delayed events and send the join state event immediately
+                    if (error) {
+                        logger.info("Not using deleayed event because: " + error);
```
```suggestion
                        logger.info("Not using delayed event because: " + error);
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,859 @@
+     * It will send delayed events and membership events
+     * @param fociPreferred
+     * @param focusActive
+     * @param onError This will be called once the membership menager encounters an unrecoverable error.
```
```suggestion
     * @param onError This will be called once the membership manager encounters an unrecoverable error.
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,859 @@
+ *  - When the state machine is stopped:
+ *   - Disconnect the member
+ *   - Stop the timer for the delay refresh
+ *   - Stop the timer for updateint the state event
```
```suggestion
 *   - Stop the timer for updating the state event
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,859 @@
+    private insertions: Action[] = [];
+    private resetWith?: Action[];
+    /**
+     * This starts the main loop of the memberhsip manager that handles event sending, delayed event sending and delayed event restarting.
```
```suggestion
     * This starts the main loop of the membership manager that handles event sending, delayed event sending and delayed event restarting.
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,859 @@
+    private resetWith?: Action[];
+    /**
+     * This starts the main loop of the memberhsip manager that handles event sending, delayed event sending and delayed event restarting.
+     * @param initialActions The initial actions the manager will start with. It should be enough to pass: DelayedLeaveActionType.Initial
```
```suggestion
     * @param initialActions The initial actions the manager will start with. It should be enough to pass: DelayedLeaveActionType.Initial
     * @returns Promise that resolves once all actions have run and no more are scheduled.
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,859 @@
+/**
+ * This class takes care of the membership management.
+ * It has the following tasks:
+ *  - Send the users leave delayed event before sending the memberhsip
```
```suggestion
 *  - Send the users leave delayed event before sending the membership
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,859 @@
+                                return;
+                            }
+                            if (this.unsupportedDelayedEndpoint(e)) {
+                                logger.info("Not using deleayed event because the endpoint is not supported");
```
```suggestion
                                logger.info("Not using delayed event because the endpoint is not supported");
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,907 @@
+            error.httpStatus < 600
+        ) {
+            logger.warn(
+                "Network error while sending event, retrying in " + retryDurationString + " " + retryCounterString,
```
```suggestion
                "Server error while sending event, retrying in " + retryDurationString + " " + retryCounterString,
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,907 @@
+                            if (this.retryOnNetworkError(e, type)) return;
+
+                            // On any other error we fall back to SendLeaveEvent (this includes hard errors from rate limiting)
+                            this.scheduler.addAction({ ts: Date.now(), type: MembershipActionType.SendLeaveEvent });
```
I'm not sure if `info` or `warn` is appropriate 🤷

hughns reviewed (commented):
As per DM this is looking much saner now the network error logic is in place. I will have time to give it a proper review tomorrow.

dbkr reviewed (commented):
Phew, this is a big one. Also, nitpick: newlines between functions / tests / any individual thing that spans more than one line (like a variable declaration with a separate line comment) are quite nice.

toger5 commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,896 @@
+ * It can also be awakened whenever a new action is added which is
+ * earlier then the current "next awake".
+ * @internal
+ */
```
> What advantage does this have over just setting a separate timer for each thing?

With seperate timers we can always run the rist to have parallel loops. This is an easy way to track what else is schduled.

> I know this is private but it really feels like it needs its own set of tests. It's probably big enough to warrant its own file anyway.

It started as part of the class itself. I split it out to make clear they are seperate bits of logic to understand but its not thought of a generic handler.

toger5 replied:
Addressed the undstandability by numerous changes.

toger5 commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,896 @@
+            this.didWakeUp = true;
+            this.wakeup?.();
+        }
+    }
```
It should be a noop if the Scheduler is not running. `startWithActions` should have resolved hende `addActions` cannot be called anymore.

Do you think a throw or a log would be appropriate?

toger5 replied:
added a warning

toger5 commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,896 @@
+            this.didWakeUp = true;
+            this.wakeup?.();
+        }
+    }
```
maybe calling both actions is confusing. But I think the logic checks out.

`this.actions[0]?.ts` is the ts at which we already have a timeout scheduled to continue the loop.
If the smallest ts in the new actions array (`this.resetWith`) is larger then the next ts the loop still continues at the time we have to run the next action.
But if the newest item `newestTs` (smallest ts) from `this.resetWith` is smaller (`newestTs < nextTs`) waiting for the next timeout would be too late.

toger5 replied:
killing and fully restarting is an intersting idea. I will explore that.

toger5 commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,896 @@
+    }
+
+    // Scheduler:
+    private scheduler = new ActionScheduler(ActionScheduler.defaultState, this.membershipLoopHandler.bind(this));
```
I opted for bind since its shorter. But if the style is to not bind than we go with the arrow function

toger5 commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,896 @@
+                // Before we start we check if we come from a state where we have a delay id.
+                if (!state.delayId) {
+                    // Normal case without any previous delayed id.
+                    await this.client
```
I refactored it to use then+catch because it made the main action much clearer.
`client._unstable_snedDelayedStateEvent` is more prominent now compared to inside a the first line in the try block.

toger5 commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,896 @@
+            error instanceof HTTPError &&
+            typeof error.httpStatus === "number" &&
+            error.httpStatus >= 500 &&
+            error.httpStatus < 600
```
done with this block
```
else if (error instanceof ConnectionError) {
            logger.warn(
                "Network connection error while sending event, retrying in " +
                    retryDurationString +
                    " " +
                    retryCounterString,
                error,
            );
        }
```
(this is what you mean right?)

toger5 commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,896 @@
+    }
+
+    public addAction(action: Action): void {
+        this.insertions.push(action);
```
while the `membershipLoopHandler` is running the actions array should be immutable.
I think that is not taken advantage of in the current setup. But any logic that would be based on what is scheduled next would have the guarantee, that the actions are the same during one iteration.
What is helpful however is that reset and add action do not collide and are not dependent in which order they are called.

I am sure it can also work without this but it would make it harder to predict the outcome if the order makes a difference as well.
And I think its not a lot of added code.
Wdyt?

toger5 replied:
Solved by using a return object for the new actions.

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,918 @@
+        const retries = this.scheduler.state.networkErrorRetries.get(type) ?? 0;
+        const retryDurationString = this.callMemberEventRetryDelayMinimum / 1000 + "s";
+        const retryCounterString = "(" + retries + "/" + this.maximumNetworkErrorRetryCount + ")";
+        if (error instanceof Error && error.name === "AbortError") {
```
As per DM, I don't think this will detect network errors when EC is running as embedded widget.

I *think* we would see `MatrixError`s instead which we would need to parse.

toger5 commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,918 @@
+     */
+    private rateLimitErrorHandler(error: unknown, method: string, type: MembershipActionType): boolean {
+        // "Is rate limit"-boundary
+        if (!(error instanceof HTTPError && error.isRateLimitError())) {
```
Adressed in: https://github.com/matrix-org/matrix-js-sdk/pull/4726/commits/79abd60ef8f5eb8628bcfb1adaee38e358cac3a6

hughns commented on src/embedded.ts:
```diff
@@ -673,3 +689,12 @@ function processAndThrow(error: unknown): never {
     }
 }
+
+function timeoutToConnectionError(error: unknown): never {
```
```suggestion
/**
 * This converts an "Request timed out" error from the PostmessageTransport into a ConnectionError.
 * It either throws the original error or a new ConnectionError.
 **/
function timeoutToConnectionError(error: unknown): never {
```

toger5 commented:
I think this is ready in its first iteration.
@dbkr It does not yet address your readability/understand-ability concerns.

I created this actionable issue to work through the good comments which I have not addressed yet:
https://github.com/matrix-org/matrix-js-sdk/issues/4743

The reason those get not addressed here is, that those are better approached not one by one but with a general top to bottom discussion how to make this as easy to understand as possible. (There is a chance we can drop the `SendMainDelayedEvent` action which would greatly simplify it and allow for more radical changes that can incorporate the unaddressed feedback)

dbkr commented on .github/workflows/sonarcloud.yml:
```diff
@@ -27,7 +27,7 @@ jobs:
             # We create the status here and then update it to success/failure in the `report` stage
             # This provides an easy link to this workflow_run from the PR before Sonarcloud is done.
-            - uses: guibranco/github-status-action-v2@5ef6e175c333bc629f3718b083c8a2ff6e0bbfbc
+            - uses: guibranco/github-status-action-v2@7ca807c2ba3401be532d29a876b93262108099fb
```
Possibly a merge has gone wrong here and I think you might be unintentionally reverting something?

toger5 replied:
this has been fixed on the main branch of the PR and is only present for the v0.8.0-rc.1 branch.

dbkr reviewed (commented):
Thanks for all the fixes, looking much better. I think you may have had a merge gone wrong though as it looks like there's some reverts crept in here.

toger5 commented:
> Thanks for all the fixes, looking much better. I think you may have had a merge gone wrong though as it looks like there's some reverts crept in here.
This is probably: 0cdb07544d9926cf0855a76ca5cc7dab253bdb24 which we needed to do to make CI happy. currently EC rc-1 depends on this branch. So we will merge this with develop again before we merge this.

toger5 commented:
I created the branch in this PR https://github.com/matrix-org/matrix-js-sdk/pull/4744 to keep the the commit 0cdb07544d9926cf0855a76ca5cc7dab253bdb24 alive which we need for the EC v0.8.0-rc.1. For the actual release we will merge this PR first

richvdh commented on src/errors.ts:
```diff
@@ -52,3 +52,13 @@ export class ClientStoppedError extends Error {
     }
 }
+
+export class UnsupportedEndpointError extends Error {
```
Flyby comment: please give this class (and any other exported symbols) some [documentation]. What does it represent? How should downstream applications use it? What do the properties within it mean?

See https://github.com/element-hq/element-web/blob/develop/code_style.md#comments

andybalaam commented on spec/unit/matrixrtc/MembershipManager.spec.ts:
```diff
@@ -603,4 +616,87 @@ describe.each([
     });
+    describe("unrecoverable errors", () => {
+        // !FailsForLegacy because legacy does not have a retry limit and no mechanism to communicate unrecoverable errors.
+        it("throws, when reaching maximum number of retires for initial delayed event creation !FailsForLegacy", async () => {
```
```suggestion
        it("throws, when reaching maximum number of retries for initial delayed event creation !FailsForLegacy", async () => {
```

andybalaam commented on spec/unit/matrixrtc/MembershipManager.spec.ts:
```diff
@@ -603,4 +616,87 @@ describe.each([
+            expect(delayEventSendError).toHaveBeenCalled();
+        });
+        // !FailsForLegacy because legacy does not have a retry limit and no mechanism to communicate unrecoverable errors.
+        it("throws, when reaching maximum number of retires !FailsForLegacy", async () => {
```
```suggestion
        it("throws, when reaching maximum number of retries !FailsForLegacy", async () => {
```

andybalaam commented on spec/unit/matrixrtc/MembershipManager.spec.ts:
```diff
@@ -603,4 +616,87 @@ describe.each([
+            manager.join([focus], focusActive, unrecoverableError);
+            await jest.advanceTimersByTimeAsync(1);
+
+            expect(unrecoverableError).not.toHaveBeenCalledWith();
```
Should this be `not.toHaveBeenCalled()` ?

andybalaam reviewed (approved):
This is way too huge for me to have an opinion on generally, but nothing concerning from a crypto point of view.

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -172,7 +202,7 @@ export class MembershipManager implements IMembershipManager {
-        this.scheduler.resetState();
+        this.state = MembershipManager.defaultState;
 
         this.scheduler
```
I suspect the `.then()` needs to go before `.catch()`.

toger5 replied:
ah es lint is what is making this a hard error?

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -276,9 +306,23 @@ export class MembershipManager implements IMembershipManager {
 
-    // Membership Event parameters:
+    // MembershipManager mutable state.
+    public state: ActionSchedulerState;
```
Does this need to be public?

```suggestion
    private state: ActionSchedulerState;
```

toger5 replied:
I was considering using this to simplify the useECConnection state hook in EC

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -276,9 +306,23 @@ export class MembershipManager implements IMembershipManager {
-    // Membership Event parameters:
+    // MembershipManager mutable state.
+    public state: ActionSchedulerState;
+    public static get defaultState(): ActionSchedulerState {
```
Ditto.

```suggestion
    private static get defaultState(): ActionSchedulerState {
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,945 @@
+    private oldStatus?: Status;
+    private scheduler = new ActionScheduler((type): Promise<ActionUpdate> => {
+        if (this.oldStatus) {
+            // we put this at the beginngin of the actsion scheduler loop handle callback since it is a loop this
```
```suggestion
            // we put this at the beginning of the actions scheduler loop handle callback since it is a loop this
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,945 @@
+                } else {
+                    // This can happen if someone else (or another client) removes our own membership event.
+                    // It will trigger `onRTCSessionMemberUpdate` queue `MembershipActionType.SendFirstDelayedEvent`.
+                    // We might still have our delyed event from the previous participation and dependent on the serve this might not
```
```suggestion
                    // We might still have our delayed event from the previous participation and dependent on the server this might not
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,945 @@
+                    //
+                    // In this block we will try to cancel this delayed event before setting up a new one.
+
+                    return this.cancelKnownDelayIdBeforeSendFirstDelayedEvent(type, this.state.delayId);
```
The are a few places where the `type` argument is passed, but it appears to be redundant as we only ever call it with a single value `MembershipActionType.SendFirstDelayedEvent`.

Can the type argument be removed?

```suggestion
                    return this.cancelKnownDelayIdBeforeSendFirstDelayedEvent(this.state.delayId);
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,945 @@
+                // This becomes an unhandle-able error case since sth is signifciantly off if we dont hit any of the above cases
+                // when state.delayId !== undefined
+                // We do not use ignore and log this error since we would also need to reset the delayId.
+                // It is cleaner if we the frontend rejoines instead of resetting the delayId here and behaving like in the success case.
```
```suggestion
                // It is cleaner if we, the frontend, rejoins instead of resetting the delayId here and behaving like in the success case.
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,945 @@
+                    return createReplaceActionUpdate(MembershipActionType.SendJoinEvent);
+                }
+
+                // This becomes an unhandle-able error case since sth is signifciantly off if we dont hit any of the above cases
```
```suggestion
                // This becomes an unrecoverable error case since something is significantly off if we don't hit any of the above cases
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,945 @@
+
+                // This becomes an unhandle-able error case since sth is signifciantly off if we dont hit any of the above cases
+                // when state.delayId !== undefined
+                // We do not use ignore and log this error since we would also need to reset the delayId.
```
```suggestion
                // We do not just ignore and log this error since we would also need to reset the delayId.
```

hughns commented on src/matrixrtc/NewMembershipManager.ts:
```diff
@@ -0,0 +1,945 @@
+                this.stateKey,
+            )
+            .then((response) => {
+                // Success we reset retries and set delayId.
```
```suggestion
                // On success we reset retries and set delayId.
```

hughns reviewed (commented):
I like the direction this is heading 👍

hughns reviewed (commented):
Latest changes look good

dbkr reviewed (approved):
Thanks for the changes: I think splitting up the handler function has made a big difference in particular. I think I'm happy to give this a go.

hughns reviewed (approved):
Awesome. It's great to get this over the line. 👍

hughns commented:
@toger5 one more thing that could go in this PR, but could be a separate one, is adding some comments to say what is a workaround due to https://github.com/element-hq/synapse/pull/17810 having not landed.

This would make it easier to come back and remove the workaround later.

toger5 commented:
> @toger5 one more thing that could go in this PR, but could be a separate one, is adding some comments to say what is a workaround due to https://github.com/element-hq/synapse/pull/17810 having not landed.

I will do this as a seperate PR.
